### PR TITLE
WIP prow: cherrypicker: add support of external orgs/repos

### LINKED ISSUE #7587: prow: cherrypicker: allow picks across forks
I would like to be able to cherry-pick commit `A` from repo `a` into a fork of repo `a`, not just into other branches of `a`.

/kind feature
/area prow
/cc @kargakis @deads2k

### PR DESCRIPTION
**Pull request by droslean**
closes #7587

### CODE REVIEW

File: prow/external-plugins/cherrypicker/server.go [hunk 1]
```diff
@@ -147,8 +147,32 @@ func (s *Server) handleIssueComment(l *logrus.Entry, ic github.IssueCommentEvent
...
+	cherryPickMatches := cherryPickRe.FindAllStringSubmatch(ic.Comment.Body, -1)
+	if len(cherryPickMatches) == 0 || len(cherryPickMatches[0]) != 2 {
+		return nil
+	}
+
 	org := ic.Repo.Owner.Login
 	repo := ic.Repo.Name
+	targetBranch := strings.TrimSpace(cherryPickMatches[0][1])
+
+	targets, err := parseCommentArguments(strings.TrimSpace(cherryPickMatches[0][1]))
```
Comment (bbguimaraes) [inline prow/external-plugins/cherrypicker/server.go]: You can use `targetBranch`.

File: prow/external-plugins/cherrypicker/server.go [hunk 2]
```diff
@@ -492,3 +510,32 @@ func (s *Server) getPatch(org, repo, targetBranch string, num int) (string, erro
...
+		matchOrgRepo := beforeHash.FindStringSubmatch(args)
+		matchBranch := afterHash.FindStringSubmatch(args)
+
+		branch = strings.TrimSpace(matchBranch[1])
+		org = strings.TrimSpace(strings.Split(matchOrgRepo[1], "/")[0])
+		repo = strings.TrimSpace(strings.Split(matchOrgRepo[1], "/")[1])
+
+	} else if regexp.MustCompile("(.*)#(.*)").MatchString(args) {
+		repo = strings.TrimSpace(beforeHash.FindStringSubmatch(args)[1])
+		branch = strings.TrimSpace(afterHash.FindStringSubmatch(args)[1])
```
Comment (bbguimaraes) [inline prow/external-plugins/cherrypicker/server.go]: The regexp submatches can be used to simplify these two blocks.

File: prow/external-plugins/cherrypicker/server.go [hunk 3]
```diff
@@ -492,3 +510,32 @@ func (s *Server) getPatch(org, repo, targetBranch string, num int) (string, erro
...
+
+	} else if regexp.MustCompile("^[A-Za-z]+$").MatchString(args) {
+		branch = strings.TrimSpace(args)
```
Comment (bbguimaraes) [inline prow/external-plugins/cherrypicker/server.go]: `TrimSpace` is redundant, the regexp will not match strings with whitespace.

File: prow/external-plugins/cherrypicker/server_test.go [hunk 4]
```diff
@@ -375,3 +376,29 @@ func TestCherryPickPR(t *testing.T) {
...
+		{"branch": "branch/name"},
+		{"branch": "master"},
+		{"org": "org123", "repo": "repo123", "branch": "branch12345"},
+		{},
+	}
+
+	for num, testCase := range testCases {
+		targets, _ := parseCommentArguments(testCase)
+		if !reflect.DeepEqual(targets, testCasesExpected[num]) {
+			t.Fatalf("Expected %s but found %s ", testCasesExpected[num], targets)
```
Comment (stevekuznetsov) [inline prow/external-plugins/cherrypicker/server_test.go]: Use `t.Errorf` so we get all failures in one test run

File: prow/external-plugins/cherrypicker/server.go [hunk 5]
```diff
@@ -177,7 +194,7 @@ func (s *Server) handleIssueComment(l *logrus.Entry, ic github.IssueCommentEvent
 				return s.ghc.CreateComment(org, repo, num, plugins.FormatICResponse(ic.Comment, resp))
 			}
 		}
-		resp := fmt.Sprintf("once the present PR merges, I will cherry-pick it on top of %s in a new PR and assign it to you.", targetBranch)
+		resp := fmt.Sprintf("once the present PR merges, I will cherry-pick it on top of %s/%s#%s in a new PR and assign it to you.", org, repo, targetBranch)
```
Comment (stevekuznetsov) [inline prow/external-plugins/cherrypicker/server.go]: You need a reverse formatter:

```go
func formatTarget(org, repo, branch string) string {
	var target string
	if org != "" {
		target += org
		target += "/"
	}
	if repo != "" {
		target += repo
		target += "#"
	}
	target += branch
	return target
}
```

File: prow/external-plugins/cherrypicker/server.go [hunk 6]
```diff
@@ -492,3 +509,27 @@ func (s *Server) getPatch(org, repo, targetBranch string, num int) (string, erro
...
+		targets["repo"] = strings.TrimSpace(strings.Split(matchOrgRepo[1], "/")[1])
+
+	} else if regexp.MustCompile("(.*)#(.*)").MatchString(args) {
+		targets["repo"] = strings.TrimSpace(regexp.MustCompile("(.*)#").FindStringSubmatch(args)[1])
+		targets["branch"] = strings.TrimSpace(regexp.MustCompile("#(.*)").FindStringSubmatch(args)[1])
+
+	} else if regexp.MustCompile("[a-zA-Z0-9]").MatchString(args) {
+		targets["branch"] = strings.TrimSpace(args)
+	} else {
+		err = fmt.Errorf("Parsing error: %s\n", args)
```
Comment (stevekuznetsov) [inline prow/external-plugins/cherrypicker/server.go]: This is not going to be a useful error for the user

File: prow/external-plugins/cherrypicker/server.go [hunk 7]
```diff
@@ -492,3 +509,27 @@ func (s *Server) getPatch(org, repo, targetBranch string, num int) (string, erro
 func normalize(input string) string {
 	return strings.Replace(input, "/", "-", -1)
 }
+
+func parseCommentArguments(args string) (map[string]string, error) {
```
Comment (stevekuznetsov) [inline prow/external-plugins/cherrypicker/server.go]: GitHub ate my comment .... but lets use `strings.Contains()` to check for `#` and then for `/` and not use regex here. My sample impl for you was 10 lines

**Comment by kargakis** [comment]
Needs to be implemented for pull_request events too

File: prow/external-plugins/cherrypicker/server.go (hunk 5 above)
Comment (droslean) [reply prow/external-plugins/cherrypicker/server.go]: My intention here is to print all the info (org/repo#branch) regardless the request. Even if the user requested to cherrypick only a branch, still those values won't be empty.

**Comment by kargakis** [comment]
reopen if you plan to rework this

/close

**Pull request closed without merging**